homeassistant/components/nina/binary_sensor.py

```diff
@@ -0,0 +1,179 @@
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import (
+    _LOGGER,
+    ATTR_EXPIRES,
+    ATTR_HEADLINE,
+    ATTR_ID,
+    ATTR_SENT,
+    ATTR_START,
+    CONF_FILTER_CORONA,
+    CONF_MESSAGE_SLOTS,
+    CONF_REGIONS,
+    DOMAIN,
+    SCAN_INTERVAL,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,  # pylint: disable=C0330
+    config_entry: ConfigEntry,  # pylint: disable=C0330
+    async_add_entities: Any,  # pylint: disable=C0330
+) -> None:
+    """Set up entries."""
+    config: Dict[str, Any] = hass.data[DOMAIN][config_entry.entry_id]
+
+    regions: Dict[str, str] = config[CONF_REGIONS]
+
+    filter_corona: bool = config[CONF_FILTER_CORONA]
+
+    nina: Nina = Nina(async_get_clientsession(hass))
+
+    for region in regions.keys():
+        nina.addRegion(region)
+
+    async def async_update_data():
+        """Fetch data from NINA."""
+        try:
+            await nina.update()
+
+            return nina.warnings
+
+        except ApiError as err:
+            _LOGGER.warning("NINA connection error: %s", err)
+
+    coordinator: DataUpdateCoordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="sensor",
+        update_method=async_update_data,
+        update_interval=SCAN_INTERVAL,
+    )
+
+    await coordinator.async_config_entry_first_refresh()
+
+    entities: List[NINAMessage] = []
+
+    for idx, ent in enumerate(coordinator.data):  # pylint: disable=W0612
+        for i in range(0, config[CONF_MESSAGE_SLOTS]):
+            entities.append(
+                NINAMessage(coordinator, ent, regions[ent], i + 1, filter_corona)
+            )
+
+    async_add_entities(entities)
+
+
+class NINAMessage(CoordinatorEntity, BinarySensorEntity):  # pylint: disable=R0902
+    """Representation of an NINA warning."""
+
+    def __init__(  # pylint: disable=R0913
+        self,  # pylint: disable=C0330
+        coordinator: DataUpdateCoordinator,  # pylint: disable=C0330
+        region: str,  # pylint: disable=C0330
+        regionName: str,  # pylint: disable=C0330
+        slotID: int,  # pylint: disable=C0330
+        filter_corona: bool,  # pylint: disable=C0330
+    ):
+        """Initialize."""
+        super().__init__(coordinator)
+
+        self._region: str = region
+        self._region_name: str = regionName
+        self._has_warning: bool = False
+        self._slot_id: int = slotID
+
+        self._message_id: str = ""
+        self._message_heading: str = ""
+        self._message_sent: str = ""
+        self._message_start: str = ""
+        self._message_expires: str = ""
+
+        self._filter_corona: bool = filter_corona
+
+        self._coordinator: DataUpdateCoordinator = coordinator
+
+        self.parse_update()
+
+    def parse_update(self):
```

Comment:
Maybe implement a custom `DataUpdateCoordinator` and take care of the data mapping there?

Because calling `self.parse_update()` before every operation isn't really nice.